### PR TITLE
chore: migrate more codebase to ESM

### PR DESCRIPTION
**Pull request by danez**
#### Summary

Migrates some parts of the codebase to ESM.

To move further I have to first introduce vitest for the unit tests to be able to properly mock modules, which rewiremock fails with esm now.

---

For us to review and ship your PR efficiently, please perform the following steps:

- [ ] Open a [bug/issue](https://github.com/netlify/cli/issues/new/choose) before writing your code 🧑‍💻. This ensures we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [ ] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and
      passes our tests.
- [ ] Update or add tests (if any source code was changed or added) 🧪
- [ ] Update or add documentation (if features were changed or added) 📝
- [ ] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**

### CODE REVIEW

File: bin/run.mjs [hunk 1]
```diff
@@ -1,15 +1,14 @@
...
 
 import updateNotifier from 'update-notifier'
 
 import { createMainCommand } from '../src/commands/index.mjs'
+import getPackageJson from '../src/utils/get-package-json.mjs'
 
 // 12 hours
 const UPDATE_CHECK_INTERVAL = 432e5
-const pkg = JSON.parse(readFileSync(fileURLToPath(new URL('../package.json', import.meta.url))), 'utf-8')
+const pkg = await getPackageJson()
```
Comment (danez) [inline bin/run.mjs:11]: This was extracted into a separate file, so we do not have to do this all over again multiple times.

File: src/commands/base-command.mjs [hunk 2]
```diff
@@ -8,29 +8,25 @@ import debug from 'debug'
...
   pollForToken,
   sortOptions,
-  track,
   warn,
-} = utils
+} from '../utils/command-helpers.cjs'
+import getGlobalConfig from '../utils/get-global-config.cjs'
+import { openBrowser } from '../utils/open-browser.cjs'
+import StateConfig from '../utils/state-config.mjs'
+import { identify, track } from '../utils/telemetry/index.mjs'
```
Comment (danez) [inline src/commands/base-command.mjs:29]: I started unusing `/utils/index.cjs` :

1. It is easier to migrate
2. It makes node not require everything at once, which might come in handy once the commands are not all loaded always, and only the command in use.

File: tests/unit/utils/functions/get-functions.test.mjs [hunk 3]
```diff
@@ -78,36 +78,3 @@ test('should mark background functions based on filenames', async (t) => {
...
-        name: 'index',
-        mainFile: path.join(builder.directory, 'functions', 'index.js'),
-        isBackground: false,
-        runtime: 'js',
-        urlPath: '/.netlify/functions/index',
-      },
-    ])
-    t.deepEqual(watchDirs, [path.join(builder.directory, 'functions'), path.join(builder.directory, 'utils')])
-  })
-})
```
Comment (danez) [inline tests/unit/utils/functions/get-functions.test.mjs:113]: `getFunctionsAndWatchDirs` was unused, so removed.

File: src/utils/functions/get-functions.mjs [hunk 4]
```diff
@@ -34,23 +33,3 @@ const getFunctions = async (functionsSrcDir, config = {}) => {
...
-
-  // get all functions files so we know which directories to watch
-  const functions = await listFunctions(functionsSrcDir)
-  const watchDirs = [functionsSrcDir]
-
-  // filter for only main files to serve
-  const functionsWithProps = functions.filter(({ runtime }) => runtime === JS).map((func) => addFunctionProps(func))
-
-  return { functions: functionsWithProps, watchDirs }
-}
```
Comment (danez) [inline src/utils/functions/get-functions.mjs:54]: `getFunctionsAndWatchDirs` was unused so removed

File: src/utils/get-package-json.mjs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+let packageJson
+
+const getPackageJson = async () => {
+  if (!packageJson) {
+    const packageJsonPath = join(dirname(fileURLToPath(import.meta.url)), '../../package.json')
+    packageJson = JSON.parse(await readFile(packageJsonPath))
+  }
+
+  return packageJson
```
Comment (danez) [inline src/utils/get-package-json.mjs:13]: This now also caches the root `package.json` of the cli as I do not think it will change during runtime.

**Review by khendrikse — APPROVED** [review]

**Pull request merged**